freeipa-api/src/main/java/com/sequenceiq/freeipa/api/v1/freeipa/stack/FreeIpaV1Endpoint.java

```diff
@@ -185,6 +188,18 @@
     @PUT
     @Path("salt_update")
     @Produces(MediaType.APPLICATION_JSON)
-    @ApiOperation(value = UPDATE_SALT, nickname = "updateSaltV1")
+    @ApiOperation(value = UPDATE_SALT, produces = MediaType.APPLICATION_JSON, notes = FreeIpaNotes.FREEIPA_NOTES, nickname = "updateSaltV1")
     FlowIdentifier updateSaltByName(@QueryParam("environment") @NotEmpty String environmentCrn, @AccountId @QueryParam("accountId") String accountId);
+
+    @PUT
+    @Path("retry")
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(value = RETRY, produces = MediaType.APPLICATION_JSON, notes = FreeIpaNotes.FREEIPA_NOTES, nickname = "retryV1")
+    FlowIdentifier retry(@QueryParam("environment") @NotEmpty String environmentCrn);
+
+    @GET
+    @Path("retry")
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(value = LIST_RETRYABLE_FLOWS,  produces = MediaType.APPLICATION_JSON, notes = FreeIpaNotes.FREEIPA_NOTES, nickname = "listRetryableFlowsV1")
+    List<RetryableFlowResponse> listRetryableFlows(@QueryParam("environment") @NotEmpty String environmentCrn);
```

Comment:
It is not clear to me what service is intended to call this and then invoke the retry methods. Previously I thought FMS handled FMS retries. But this looks like it is setting up for a different service to trigger retries. I am just curious.

Reply:
you could retry upgrade if it fails at some point. this is implemented like we have for clusters. UI could query this to see if there anything to retry and decide to show the button or not